ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/RootPeersDNS.hs

```diff
@@ -642,6 +656,52 @@ prop_local_updatesDomainsCorrectly mockRoots@(MockRoots lrp _ _ _)
      in property (fst r)
 
 
+-- | The 'localRootPeersProvider' monitors each domain address for DNS
+-- resolving. It then should update the local result group correctly, but it
+-- should only update the single source of truth and not some cached version.
+--
+-- This test follows #3638 and checks this doesn't happen.
+--
+-- The way we test for this is to look at TraceLocalRootGroups results and check
+-- if they are monotonic i.e. the most recent value of TraceLocalRootGroups has
+-- to be subset or equal to the last values of TraceLocalRootGroups. Since we
+-- only do fixed DNS resolution, the single source of truth should be only
+-- increasing on not decreasing.
+--
+prop_local_singleSourceOfTruth :: MockRoots
+                               -> Script DNSTimeout
+                               -> Script DNSLookupDelay
+                               -> Property
+prop_local_singleSourceOfTruth mockRoots
+                               dnsTimeoutScript
+                               dnsLookupDelayScript =
+    let tr = selectTraceEventsDynamic
+              @() @(Solo (Seq (Int, Map SockAddr PeerAdvertise)))
+           $ runSimTrace
+           $ mockLocalRootPeersProvider tracerTraceLocalRoots
+                                        mockRoots
+                                        dnsTimeoutScript
+                                        dnsLookupDelayScript
+
+     in fst
+      $ foldl'
+          (\(r, prev) t ->
+            case t of
+              Solo curr ->
+                let currSet = Set.fromList
+                            . concatMap (Map.keys . snd)
+                            . toList
+                            $ curr
+                    isMonotonic = prev `Set.isSubsetOf` currSet
```

Comment:
This is not the right property, it only works because of #3650.

Reply:
Right, but if you are saying that we can not go for a monotonicity check because older entries might be removed then it probably is impossible to distinguish a single source of truth as well, no?

How can I just by observing the contents of the TVar tell if its a SST ?

Reply:
hmmm it sounds tricky, but doable! I'll think about this!

Reply:
I have been thinking and experimenting about this and the best way to do this is once we have garbage collection we update this test to accommodate it. We should have a trace for when and which DNS entries were gc'd and then we would update our expected `localRootPeerSet` to be monotonic modulo garbage collection.